### PR TITLE
Use TryGetValue when possible

### PR DESCRIPTION
**Pull request by forki**

### CODE REVIEW

**Comment by forki** [comment]
this is ready

File: src/fsharp/TypedTreeOps.fs [hunk 1]
```diff
@@ -1393,11 +1393,13 @@ type ValHash<'T> =
         let (ValHash t) = ht
         t.Values :> seq<'T>
 
-    member ht.TryFind (v: Val) = 
+    member ht.TryGetValue(v:Val, [<System.Runtime.InteropServices.Out>] value: byref<'T>) =
```
Comment (cartermp) [inline src/fsharp/TypedTreeOps.fs:1396]: Is this attribute required, or is just the byref sufficient?

Comment (forki) [reply src/fsharp/TypedTreeOps.fs:1396]: I am not totally sure tbh. I copied it from other places. AFAIK it was required earlier

Comment (TIHan) [reply src/fsharp/TypedTreeOps.fs:1396]: This should be `TryGetValue(v: Val, value: outref<'T>)`

**Comment by KevinRansom** [comment]
@TIHan , tryget value returning a bool is a pretty decent pattern in F#.
 ````
           match NameMap.tryFind assemblyName t.CcuTable with
            | Some res -> Some res
            | None -> 
`````
As opposed to this is about equivalent readability.  In C# prior to 8.0 trygetvalue was not great at all...:
````
            match t.CcuTable.TryGetValue assemblyName with
            | true, res -> ValueSome res
            | _ -> 
````

**Comment by TIHan** [comment]
@KevinRansom It is decent and common, but TryFind is simply better except when it becomes a perf issue, which in this case, no evidence has been shown. If having changed everything to TryGetValue gave us no perf benefit, it would just mean that we have less pretty code to look at it.

**Comment by TIHan** [comment]
I may have an idea where we could get TryFind with the return option to be just as performant as TryGetValue, effectively it would emit a TryGetValue in the end. We wouldn't just change an expression tree of a TryFind call to a TryGetValue. We would rely on a particular expression pattern to determine if an option could be eliminated; would require some work but it is possible; kinda of like detupling.

**Comment by forki** [comment]
TryFind perf issues where shown over and over again in the past.

Will Smith <notifications@github.com> schrieb am Mi., 27. Mai 2020, 03:26:

**Comment by KevinRansom** [comment]
@TIHan , here is my view:  https://github.com/dotnet/fsharp/pull/9286#issuecomment-635123464

**Comment by TIHan** [comment]
> TryFind perf issues where shown over and over again in the past.

I understand TryFind has perf issues when called a large number of times. But in the cases you have provided, are they actually causing problems? Are you able to show a trace with the evidence? Or a benchmark?

I know this might be annoying, but it's important we have the information. If it were me, I wouldn't make this kind of PR unless I was able to measure or have a trace of evidence. I want the evidence because the code is not as idiomatic as TryFind because of the use of `byref/outref` and in-ability to use the `Option` API. Had the code be just as idiomatic I would be more inclined to accept the PR even if there is little evidence.

**Comment by forki** [comment]
TryGetValue is super idiomatic in dotnet and F#. Those calls are in hot
paths during type checking and replacing them with voption and byref is
working like dominoes here

Will Smith <notifications@github.com> schrieb am Fr., 29. Mai 2020, 01:51:

**Comment by TIHan** [comment]
@forki , are you able to provide evidence?

**Comment by dsyme** [comment]
Hmmm I read the discussion more carefully.

I personally have grown used to TryGetValue and immediate matching and I'm ok with that as basically idomatic in the compiler if a lookup supports TryGetValue.

I'm much less ok with the peppering of ValueNone/ValueSOme through the codebase.

**Comment by KevinRansom** [comment]
@forki ,
thanks for this PR, I am closing this because despite the probable benefit of eliminating a few allocations, it does sit in an allocation heavy part of the codebase and so the benefits will likely be drowned out by other allocation noise.  I think once we have added some sugar to make value options less jarring to use we can revisit these types of opportunistic optimizations.

Thanks again for this and all of your other work that benefits F#

Kevin

**Pull request closed without merging**